### PR TITLE
Add simplified sparse attention algorithm over GQA to the hybrid model

### PR DESCRIPTION
**Pull request by alokpathy**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

This PR includes a simplified sparse attention indexer. A summary of the changes included in  summary of the changes in each file

| File | Change |
|---|---|
| `megatron/core/transformer/experimental_attention_variant/dsa_gqa.py` | **New.** Simplified indexer over GQA: `DSGroupedSelfAttention`, `DSGQACoreAttention`, the simplified indexer, the PyTorch reference sparse-attention paths (dense-mask and gather), and the indexer KL loss. |
| `megatron/core/transformer/experimental_attention_variant/dsa_min_memory_triton.py` | **New.** Triton kernels for the simplified indexer: selected-score forward/backward, gathered linear wgrad, RMSNorm stats, top-k index blocks. |
| `megatron/core/transformer/experimental_attention_variant/dsa_min_memory.py` | **New.** Minimum-activation memory autograd functions over those kernels (`DSASimplifiedMinMemoryGQAFn`, `DSASimplifiedDenseIndexerLossFn`), tile-wise projection/routing helpers, per-stage timing profiler. |
| `megatron/core/transformer/experimental_attention_variant/dsa_layer_specs.py` | **New.** `dsa_stack_spec`, derived from upstream's `hybrid_stack_spec`. |
| `megatron/core/transformer/experimental_attention_variant/dsa.py` | Query and key chunked top-k helpers (`fused_qk_topk_chunked`, `_merge_topk_scores`), and indexer loss |
| `megatron/core/transformer/attention.py` | Two base-class hooks: `_get_core_attention_extra_kwargs` (pass extra inputs into core attention) and `_dynamic_core_attention_forward` (raises by default so DSA + dynamic batching fails loudly instead of silently running dense). Needed to pass indexer parameters into Attention forward. |
| `megatron/core/transformer/transformer_config.py` | 31 `dsa_*` fields and validation. Adds six rope fields to base config because the GQA indexer builds its own `RotaryEmbedding` from `config`. |
| `megatron/core/optimizer/optimizer.py` | `is_dsa_indexer` param-group identifier; separate grad-norm computation and clipping for indexer vs backbone. |
| `megatron/core/optimizer/distrib_optimizer.py` | Per-parameter-group optimizer step clocks, so a reset indexer restarts at step 0 while the backbone continues. Required by `--dsa-reset-indexer-on-load`. |
| `megatron/core/optimizer/optimizer_config.py` | `dsa_separate_indexer_grad_clip`, `dsa_indexer_clip_grad`. |
| `megatron/core/optimizer_param_scheduler.py` | Indexer learning-rate logging (`get_indexer_lr_for_logging`). |
| `megatron/core/optimizer/__init__.py` | Param-group plumbing for the indexer split. |
| `megatron/training/training.py` | Parameter freezing for `--dsa-train-indexer-only` / `--dsa-train-main-only`, indexer reset-on-load, indexer LR warmup, and split grad-norm/zeros logging. Parameter freezing is necessary for multistage long-context extension. |
| `megatron/training/checkpointing.py` | Restores model-defining DSA args; asserts trainability mode is unchanged when loading optimizer state |
| `megatron/training/arguments.py` | 32 `--dsa-*` flags and their validation. |
| `megatron/training/models/dist_utils.py` | Calls `apply_dsa_param_freezing` before DDP wrapping |
| `megatron/training/utils/common_utils.py` | Split indexer/backbone gradient norms and zero counts for TensorBoard/wandb. |
| `megatron/training/utils/__init__.py` | Re-exports those helpers. |
| `tests/unit_tests/transformer/test_attention_variant_dsa_gqa.py` | **New.** ~90 tests: numerical parity against the PyTorch reference, Triton kernel checks, config validation, checkpoint/reset, freeze behavior. |
| `tests/unit_tests/test_optimizer.py` | Indexer param-group separation, split clipping, per-group step round-trip. |
| `tests/unit_tests/test_optimizer_param_scheduler.py` | Indexer LR scheduling. |
:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/arguments.py [hunk 1]
```diff
@@ -1030,6 +1029,111 @@ def validate_args(args, defaults={}):
             '--overlap-param-gather only supported with distributed optimizer, megatron fsdp, or dist_muon'
         assert args.overlap_grad_reduce, \
             'Must use --overlap-param-gather with --overlap-grad-reduce'
+        assert not args.use_legacy_models, \
+            '--overlap-param-gather only supported with MCore models'
```
Comment (Mellonta) [inline megatron/training/arguments.py:1046]: --use-legacy-models was removed from main.

File: megatron/training/checkpointing.py [hunk 2]
```diff
@@ -204,6 +206,31 @@ def _compare(arg_name, old_arg_name=None, default=None):
...
     _compare('add_position_embedding', default=True)
+    _compare('experimental_attention_variant', default=None)
+    _compare('dsa_indexer_mode', default='standard')
+    _compare('dsa_simplified_use_learned_k', default=False)
+    _compare('dsa_simplified_indexer_disable_main_input_norm', default=False)
+    _compare('dsa_standard_indexer_use_main_input_norm', default=False)
+    _compare('dsa_indexer_n_heads', default=None)
+    _compare('dsa_indexer_head_dim', default=None)
+    _compare('dsa_indexer_topk', default=None)
+    _compare('dsa_indexer_use_hadamard', default=False)
```
Comment (Mellonta) [inline megatron/training/checkpointing.py:217]: How does it load a existing pre-training checkpoint?

Comment (alokpathy) [reply megatron/training/checkpointing.py:217]: The callsite is gated with `not args.finetune`. Would loading a pretrained checkpoint go down this path, then?

File: megatron/training/arguments.py (hunk 1 above)
Comment (alokpathy) [reply megatron/training/arguments.py]: fixed, thanks!

File: megatron/core/optimizer/__init__.py [hunk 3]
```diff
@@ -351,12 +351,13 @@ def _get_param_groups(
...
 
             is_expert_parallel = not getattr(param, 'allreduce', True)
+            is_dsa_indexer = name.startswith("indexer.") or ".indexer." in name
 
             # Create config_tuple that is hash-able, and has a consistent ordering of the keys.
             param_override_tuple: tuple[tuple[str, Any], ...] | None = (
                 param_group_override_to_tuple(param_override)
             )
-            key = (param_override_tuple, is_expert_parallel)
+            key = (param_override_tuple, is_expert_parallel, is_dsa_indexer)
```
Comment (mkhona-nvidia) [inline megatron/core/optimizer/__init__.py:360]: @deepakn94  how do we deal with this list expanding the same way as previous things kept expanding?

File: megatron/core/optimizer/__init__.py [hunk 4]
```diff
@@ -373,9 +374,16 @@ def _get_param_groups(
...
-    for key in sorted(params_key, key=lambda x: (x[0] is not None, x[0])):
-        param_override_tuple, is_expert_parallel = key
+    # Keys gathered from other ranks or module domains may predate the is_dsa_indexer
+    # element; treat a missing element as False so ordering stays total either way.
+    def _key_order(key):
+        return (key[0] is not None, key[0], key[1], key[2] if len(key) > 2 else False)
+
+    for key in sorted(params_key, key=_key_order):
+        param_override_tuple, is_expert_parallel = key[0], key[1]
+        is_dsa_indexer = key[2] if len(key) > 2 else False
```
Comment (mkhona-nvidia) [inline megatron/core/optimizer/__init__.py:386]: @deepakn94  Leads to things such as this line above

Comment (mkhona-nvidia) [reply megatron/core/optimizer/__init__.py:386]: @alokpathy  if possible, could we convert this `key` variable from a tuple to a dict? and then index into the dictionary with it's appropriate key rather than index? If/when we add more stuff to the `key` it will be easier to manage. Will avoid hacky stuff like `_key_order` with `.get(,False)`

**Comment by guihong-nv** [comment]
@alokpathy could you add some comments to clarify if the added kernel is deterministic or not?
